librad/src/git/refs.rs

```diff
@@ -214,29 +229,44 @@ impl Refs {
     pub fn compute(storage: &Storage, urn: &Urn) -> Result<Self, stored::Error> {
         let namespace = Namespace::from(urn);
         let namespace_prefix = format!("refs/namespaces/{}/", namespace);
-        let heads_ref = Reference::heads(namespace, None);
-
-        tracing::debug!("reading heads from {}", &heads_ref);
 
-        let heads = storage
-            .references(&heads_ref)?
-            // FIXME: this is `git_ext::reference::iter::References::peeled()`,
-            // which we need to generalise to allow impl Iterator combinators
-            .filter_map(|reference| {
-                reference.ok().and_then(|head| {
-                    head.name()
-                        .and_then(|name| head.target().map(|target| (name.to_owned(), target)))
-                })
+        fn peeled(r: Result<git2::Reference, storage::Error>) -> Option<(String, git2::Oid)> {
+            r.ok().and_then(|head| {
+                head.name()
+                    .and_then(|name| head.target().map(|target| (name.to_owned(), target)))
             })
-            .try_fold(BTreeMap::new(), |mut acc, (name, oid)| {
-                tracing::trace!("raw refname: {}", name);
-                let name = name.strip_prefix(&namespace_prefix).unwrap_or(&name);
-                tracing::trace!("stripped namespace: {}", name);
-                let refname = reference::RefLike::try_from(name)?;
-                acc.insert(reference::OneLevel::from(refname), oid.into());
+        }
+
+        let refined = |(name, oid): (String, git2::Oid)|
+             -> Result<(reference::OneLevel, Oid), stored::Error>
+        {
+            let name = reference::RefLike::try_from(
+                name.strip_prefix(&namespace_prefix).unwrap_or(&name)
+            )?;
+            Ok((reference::OneLevel::from(name), oid.into()))
+        };
 
-                Ok::<_, stored::Error>(acc)
-            })?;
+        let heads = storage
+            .references(&Reference::heads(namespace.clone(), None))?
+            .filter_map(peeled)
+            .map(refined)
+            .collect::<Result<_, _>>()?;
+        let rad = storage
+            .references(&Reference::rads(namespace.clone(), None))?
+            .filter_map(peeled)
+            .filter(|(name, _)| !name.ends_with("rad/signed_refs"))
```

Comment:
I'm unsure if this should also filter out `rad/id`. `rad/ids/*` :thinking:

Reply:
What's the usage of `Refs`? How useful is it for those to be included or would always just access them by building the reference ourselves, i.e. `Reference::rad_id`, `Reference::rad_ids`?

Or in another light, what's the purpose of the `rad/*` references in here? Is it for the extension of new application level branches, e.g. the revisions?

Reply:
We only ever fetch the refs of peer `A` which are listed in the `signed_refs`, which must be signed by `A`, and where the Oids match the advertised ones. Recall: we could be fetching `A`s refs from untrusted `B` (ie. `refs/remotes/A/*`), so if we didn't do this, `B` could give us anything.

I'm including `rad/*` here because if we want to namespace "special" refs there, they need to get replicated. Clearly, we can't include `rad/signed_refs`, because the hash will change after we committed the result of `compute`. For everything else, the question is would it interfere with anything? We fetch `rad/id` and `rad/ids/*` first (without looking at `rad/signed_refs`), so it should not have any effect. _Strictly speaking_, we could omit them, thus. But I'm wondering if it would make things more robust against future changes if we have less exceptions from the rule.

Reply:
> But I'm wondering if it would make things more robust against future changes if we have less exceptions from the rule.

I'm gravitating towards this sentiment more. Let's not have edge-cases that we will have to keep track of and workaround. It's easier to filter any unwanted `rad` refs downstream than it is to add them back in.

Reply:
For correctness, we would need to trigger a `Refs::update` after messing with the identity, tho :thinking: